### ISSUE TITLE
Uploading a file with multipart not working properly

### ISSUE THREAD

**Issue by fredo838** [omarryhan/aiogoogle#46]
I have following code: 

```
async def async_gcs(file_list: List[Tuple[str, str, str]], bucket: str) -> List[Dict]:
    """Upload files in parallel to GCS

    Args:
        file_list (List[Tuple[str, str, str]]): list of files to upload
        bucket (str): bucket to upload to

    Returns:
        List[Dict]: list http responses payloads
    """
    async with aiogoogle_factory:
        storage = await aiogoogle_factory.discover("storage", "v1")
        requests = []
        for file_name, file_path, mime_type in file_list:
            req = storage.objects.insert(bucket=bucket, name=file_name, upload_file=file_path)
            req.headers = {'Content-Type': mime_type}
           # req.media_upload.multipart = False
            requests.append(req)
        results = await aiogoogle_factory.as_service_account(*requests, full_res=True)
    return [results.json] if len(file_list) == 1 else [result.json for result in results]

file_list = [('<destination_name>','<path_to_csv_file>', '<file_mime_type>')] # in this case file_mime_type = "text/csv"
asyncio.run(async_gcs(file_list, bucket))
```
Most files get uploaded fine, but when I try to upload a `text/csv` file, my *file content changes* to something like 
```
--1d2d26ead9e648f9a3732162b709808a
Content-Type: application/octet-stream

,text,conf,page_num,xmin,xmax,ymin,ymax,font_size,line_id,box_id,origin
0,SPORTING,0.99375,1,6.751644020536765,160.24835550262605,35.00000000000006,74.00001525878909,39,0,0,image/png
1,A,0.92,1,55.999996185302734,117.0,71.0,149.0,78,1,1,image/png
2,TICKET,0.995,1,216.00001525878906,321.0,193.0,227.0,34,2,3,image/png
3,JAARABONNEMENTEN,0.84,1,340.0,662.0,193.0,227.0,34,2,3,image/png
4,&,0.94,1,673.0,691.0,193.0,227.0,34,2,3,image/png
5,MEERBEL,0.9342857142857143,1,703.0,838.0,193.0,227.0,34,2,3,image/png
6,STEDELIJKE,0.9940000000000001,1,253.0,523.0,246.0,290.99999618530273,45,3,4,image/png
7,ZWEMBADEN,0.9855555555555555,1,533.0,825.0,246.0,290.99999618530273,45,3,4,image/png

--1d2d26ead9e648f9a3732162b709808a
Content-Type: application/json
Content-Length: 4

null
--1d2d26ead9e648f9a3732162b709808a--

```

The original content is 
```
,text,conf,page_num,xmin,xmax,ymin,ymax,font_size,line_id,box_id,origin
0,SPORTING,0.99375,1,6.751644020536765,160.24835550262605,35.00000000000006,74.00001525878909,39,0,0,image/png
1,A,0.92,1,55.999996185302734,117.0,71.0,149.0,78,1,1,image/png
2,TICKET,0.995,1,216.00001525878906,321.0,193.0,227.0,34,2,3,image/png
3,JAARABONNEMENTEN,0.84,1,340.0,662.0,193.0,227.0,34,2,3,image/png
4,&,0.94,1,673.0,691.0,193.0,227.0,34,2,3,image/png
5,MEERBEL,0.9342857142857143,1,703.0,838.0,193.0,227.0,34,2,3,image/png
6,STEDELIJKE,0.9940000000000001,1,253.0,523.0,246.0,290.99999618530273,45,3,4,image/png
7,ZWEMBADEN,0.9855555555555555,1,533.0,825.0,246.0,290.99999618530273,45,3,4,image/png

```

The issue is fixed when I uncomment the line: 
```
req.media_upload.multipart = False
```
in the first code block above. Seems like something is going on with multipart uploads that is not working correctly. Any idea what I could be doing wrong?

**Comment by omarryhan** [comment]
I found the bug that might be causing this error and just patched it in the latest commit. It was because we were not checking if there's a JSON payload before appending it to the multipart request.

Can you upgrade to v1.1.2 and let me know how it goes?

**Comment by omarryhan** [comment]
Also, it's weird that the first `part` of the request gets a Content-Type of `application/octet-stream` instead of `text/csv` or at the very least `text/plain`.

**Comment by fredo838** [comment]
With v1.1.2 I get the following file content: 
```
--5313418569ab4cc9883a1bf9ffafdb7a
Content-Type: application/octet-stream

,text,conf,page_num,xmin,xmax,ymin,ymax,font_size,line_id,box_id,origin
0,SPORTING,0.99375,1,6.751644020536765,160.24835550262605,35.00000000000006,74.00001525878909,39,0,0,image/png
1,A,0.92,1,55.999996185302734,117.0,71.0,149.0,78,1,1,image/png
2,TICKET,0.995,1,216.00001525878906,321.0,193.0,227.0,34,2,3,image/png
3,JAARABONNEMENTEN,0.84,1,340.0,662.0,193.0,227.0,34,2,3,image/png
4,&,0.94,1,673.0,691.0,193.0,227.0,34,2,3,image/png
5,MEERBEL,0.9342857142857143,1,703.0,838.0,193.0,227.0,34,2,3,image/png
6,STEDELIJKE,0.9940000000000001,1,253.0,523.0,246.0,290.99999618530273,45,3,4,image/png
7,ZWEMBADEN,0.9855555555555555,1,533.0,825.0,246.0,290.99999618530273,45,3,4,image/png

--5313418569ab4cc9883a1bf9ffafdb7a--
```
so no more "application/json", but still some weirdness

**Comment by omarryhan** [comment]
When you upload other file types, do they end up not having the problematic multipart header and footer as shown in your csv?

Also, I think what's causing this error is that the Storage API is reading the body as a simple media upload rather than a multipart upload. Can you try passing `uploadType="multipart"` and let me know how it goes?

**Comment by fredo838** [comment]
`uploadType="multipart"` yields : ```aiogoogle.excs.ValidationError: (("Invalid (extra) parameters: {'uploadType': 'multipart'} were passed",),```
in context: 
```
req = storage.objects.insert(bucket=bucket, name=file_name, upload_file=file_path, uploadType="multipart")
```

**Comment by omarryhan** [comment]
Odd. The online docs mention that this is a required parameter. Meanwhile,
in the discovery document, it's not even mentioned there.

Can you try passing `validate=false` as well?

**Comment by fredo838** [comment]
No more error, but still the same file content as before (with headers but no application/json). This back-and-forth is too slow, I created a public bucket for you to test with: `gs://aiogooogle-testing` (with 3 o's, GCP doesn't allow you to have `google` in a bucket name apparantly). Note that this bucket is unrestricted, so please don't abuse it (by uploading lot's of terrabytes, I'll have to pay for it :D). Hope it helps.

**Comment by omarryhan** [comment]
Finally, a place to store all my terabytes worth of hoarded data. Thank you :P.

I just pushed a change that lets you specify the content-type header of the file being uploaded before uploading it. So instead of the current: `Content-Type: application/octet-stream`. It should be `Content-Type: text/csv`. Hopefully Google gets smart enough and remove the multipart fluff.

I followed the guide [here](https://cloud.google.com/storage/docs/uploading-objects#rest-upload-objects) step by step.

Can you try this? (last time I promise)

Instead of

```python
req.headers = {'Content-Type': mime_type}
```

do

```python
req.upload_file_content_type = mime_type # Which should be 'text/csv' for CSV
```

If that doesn't work out, I don't think I have more tricks up my sleeve. Hopefully someone else can help us figure this out.

**Comment by omarryhan** [comment]
Oh, and also make sure to upload to v1.1.3

**Comment by omarryhan** [comment]
Since, we're now telling Google about the multipart boundary string, which in the example you posted above is: "5313418569ab4cc9883a1bf9ffafdb7a". I'm suspecting that your example might actually work without specifying a content type. Not sure though.

**Comment by fredo838** [comment]
1) With `req.upload_file_content_type = mime_type` I get:

```
--7072873459de4e75b0bee92c3411eafa
Content-Type: text/csv

,text,conf,page_num,xmin,xmax,ymin,ymax,font_size,line_id,box_id,origin
0,SPORTING,0.99375,1,6.751644020536765,160.24835550262605,35.00000000000006,74.00001525878909,39,0,0,image/png
1,A,0.92,1,55.999996185302734,117.0,71.0,149.0,78,1,1,image/png
2,TICKET,0.995,1,216.00001525878906,321.0,193.0,227.0,34,2,3,image/png
3,JAARABONNEMENTEN,0.84,1,340.0,662.0,193.0,227.0,34,2,3,image/png
4,&,0.94,1,673.0,691.0,193.0,227.0,34,2,3,image/png
5,MEERBEL,0.9342857142857143,1,703.0,838.0,193.0,227.0,34,2,3,image/png
6,STEDELIJKE,0.9940000000000001,1,253.0,523.0,246.0,290.99999618530273,45,3,4,image/png
7,ZWEMBADEN,0.9855555555555555,1,533.0,825.0,246.0,290.99999618530273,45,3,4,image/png

--7072873459de4e75b0bee92c3411eafa--
```

2) Without anything, the content is correct, but GCS doesn't recognize the mime_type correctly in the UI: 
![image](https://user-images.githubusercontent.com/11276933/104728760-b2d42f00-5737-11eb-8cf9-1f13d32112df.png)

**Comment by omarryhan** [comment]
Hey @fredo838, sorry it took me this long to fix this issue.
Turns out I was adding an invalid content-type header to the main request. Each multipart request should include both a JSON part and then the file upload part. And the parent request should only have a `multipart/related` content-type. Anyway, you shouldn't bother with any of that. I updated the upload examples, you might wanna take a look. It's much simpler now.

I'll release 2.0.1 with the fix now.
